src/test/java/ibm/jceplus/junit/base/BaseTestAES.java

```diff
@@ -74,220 +71,239 @@ public class BaseTestAES extends BaseTestCipher {
      */
     protected Cipher cp = null;
     protected boolean success = true;
-    protected int specifiedKeySize = 0;
-
-    public BaseTestAES(String providerName) {
-        super(providerName);
-        try {
-            if (warmup == false) {
-                warmup = true;
-                warmup();
-            }
-        } catch (Exception e) {
-        }
-    }
-
-    public BaseTestAES(String providerName, int keySize) throws Exception {
-        super(providerName);
-        this.specifiedKeySize = keySize;
-
-        Assume.assumeTrue(javax.crypto.Cipher.getMaxAllowedKeyLength("AES") >= keySize);
 
-        try {
-            if (warmup == false) {
-                warmup = true;
-                warmup();
-            }
-        } catch (Exception e) {
-        }
-    }
-
-    @BeforeClass
-    public static void setUpBeforeClass() throws Exception {
+    @BeforeAll
+    public void setUpBeforeClass() throws Exception {
         r.nextBytes(plainText512);
         r.nextBytes(plainText65536);
         r.nextBytes(plainText524288);
         r.nextBytes(plainText1048576);
         r.nextBytes(plainText16KB);
+
+        byte[] encodedKey128 = new byte[16];
+        byte[] encodedKey192 = new byte[24];
+        byte[] encodedKey256 = new byte[32];
+        r.nextBytes(encodedKey128);
+        r.nextBytes(encodedKey192);
+        r.nextBytes(encodedKey256);
+        key128 = new SecretKeySpec(encodedKey128, 0, encodedKey128.length, "AES");
+        key192 = new SecretKeySpec(encodedKey192, 0, encodedKey192.length, "AES");
+        key256 = new SecretKeySpec(encodedKey256, 0, encodedKey256.length, "AES");
     }
 
-    @Before
+    @BeforeEach
     public void setUp() throws Exception {
-        byte[] encodedKey = new byte[(specifiedKeySize > 0 ? specifiedKeySize : 128) / 8];
-        r.nextBytes(encodedKey);
-        key = new SecretKeySpec(encodedKey, 0, encodedKey.length, "AES");
+
+        int keySize = getKeySize();
+        if (getKeySize() == 128) {
+            key = key128;
+        } else if (keySize == 192) {
+            key = key192;
+        } else if (keySize == 256) {
+            key = key256;
+        } else {
+            throw new Exception("Illegal key size in BaseTestAES.");
+        }
     }
 
+    @Test
     public void testAES() throws Exception {
         encryptDecrypt("AES");
     }
 
+    @Test
     public void testAES_CBC_ISO10126Padding() throws Exception {
         try {
-            cp = Cipher.getInstance("AES/CBC/ISO10126Padding", providerName);
+            cp = Cipher.getInstance("AES/CBC/ISO10126Padding", getProviderName());
```

Comment:
In general i did find that using the pattern of having getter and setter methods for private fields would be easiest to maintain.

The main incentive to follow this pattern was that when tests are run they can run in any order and are likely to need values associated with a variation such as providers, key sizes, algorithm names, etc. To set these values in another copy of all the fileds in the `BaseTestJunit5` class would require reliably copying all these values ( via a single getter ) into another set of fields for the test run. Doing this in a `@BeforeEach` call would end up potentially making more calls then necessary as it would always be called before each `@Test` even if one variation in a file needs it. Using `@BeforeAll` methods can also work however i have found that if a hierarchy of tests end up having more then one @BeforeAll declared in extended classes then this ended up also causing issues running the test that might not be obvious later on. In general id expect calling getter and setter methods to have very minimal impact on overall performance and has very little downside ( id expect jit to optimize the method call out in most paths ) while makeing the code more understandable ( at least to me :) ).